File: concierge_api_client_node/authClient.ts
import * as client from 'openid-client'
import { EXPECTED_CODES } from './codes'
import { BaseConciergeClient } from './baseClient'
import { AuthzCollectionInfo } from './dataTypes'

class ConciergeTokenExpiredError extends Error {
  constructor(...params: any[]) {
    super(...params)
    this.name = 'TokenExpiredError'
  }
}

class ConciergeAuthenticationError extends Error {
  constructor(...params: any[]) {
    super(...params)
    this.name = 'ConciergeAuthenticationError'
  }
}

export class ConciergeAuthorizationClient extends BaseConciergeClient {
  token: client.TokenEndpointResponse
  openIdConfig: client.Configuration
  refreshTask: Promise<client.TokenEndpointResponse> | null = null

  constructor(serverUrl: string, token: client.TokenEndpointResponse, openIdConfig: client.Configuration) {
    super(serverUrl)
    this.token = token
    this.openIdConfig = openIdConfig
  }

  protected override async makeRequest(method: string, url: string, json?: any, files?: FormData) {
    const doRequest = async (token: client.TokenEndpointResponse) => {
      const body = (() => {
        if (json) return JSON.stringify(json)
        if (files) return files
        return undefined
      })()
      const headers = (() => {
        if (json) return {'Content-Type': 'application/json', 'Authorization': `Bearer ${token.access_token}`}
        return {'Authorization': `Bearer ${token.access_token}`}
      })()
      const response = await fetch(new URL(url, this.serverUrl), {
        method,
        body,
        headers
      })
      if (response.status === 401) {
        throw new ConciergeTokenExpiredError('Token expired')
      }
      if (response.status === 403) {
        throw new ConciergeAuthenticationError('Authentication failed')
      }
      if (!EXPECTED_CODES.includes(response.status)) {
        const text = await response.text()
        throw new Error(`Request failed with status ${response.status}: ${text}`)
      }
      return response
    }
    const currentToken = this.token
    try {
      // if we're already refreshing the token, wait for it to finish
      if (this.refreshTask) {
        await this.refreshTask
      }
      return await doRequest(this.token)
    } catch (error) {
      if (error instanceof ConciergeTokenExpiredError) {
        if (currentToken === this.token) { // this means we're probably using an expired token
          if (!this.refreshTask) {
            this.refreshTask = client.refreshTokenGrant(this.openIdConfig, this.token.refresh_token)
          }
          this.token = await this.refreshTask
          this.refreshTask = null
        } // if we're using a different token from the stored one, it's likely we already did a refresh and we can just rerun
        return await this.makeRequest(method, url, json, files)
      } else {
        throw error
      }
    }
  }

  async createCollection(collectionName: string, location: string, ownerUsername?: string): Promise<string> {
    const res = await this.makeRequest("POST", "collections", { collection_name: collectionName, location, owner_username: ownerUsername })
    const json = await res.json()
    return json.collection_id
  }

  async getCollections(): Promise<AuthzCollectionInfo[]> {
    const res = await this.makeRequest("GET", "collections")
    const json = await res.json()
    return json.map(item => new AuthzCollectionInfo(item.collection_name, item.collection_id, item.location))
  }

  async getUserInfo() {
    const res = await this.makeRequest("GET", "user_info")
    const json = await res.json()
    return json
  }

  async getPermissions() {
    const res = await this.makeRequest("GET", "permissions")
    const json = await res.json()
    return new Set(json)
  }
}

File: python_packages/shabti_api_client/src/shabti_api_client/auth_client.py
from typing import Any
import httpx
from ssl import SSLContext
from urllib.parse import urljoin
from keycloak import KeycloakOpenID
from asyncio import create_task, Task
from .exceptions import ShabtiRequestError
from .codes import EXPECTED_CODES
from shabti_types import (
    AuthzCollectionInfo,
)
from .base_client import BaseShabtiClient
from .raise_error import raise_error


class ShabtiAuthenticationError(ShabtiRequestError):
    pass


class ShabtiTokenExpiredError(ShabtiRequestError):
    pass


class ShabtiAuthorizationClient(BaseShabtiClient):
    def __init__(
        self,
        server_url: str,
        token: Any,
        keycloak_client: KeycloakOpenID,
        verify: SSLContext | None = None,
    ):
        self.server_url = server_url
        self.httpx_client = httpx.AsyncClient(verify=verify, timeout=None)
        self.token = token
        self.keycloak_client = keycloak_client
        self.refresh_task: Task | None = None

    async def _make_request(
        self, method, url, json=None, files=None, stream=False, params: dict = None
    ) -> httpx.Response:
        async def make_request(token):
            headers = (
                {"Authorization": f"Bearer {token['access_token']}"} if token else None
            )
            request = self.httpx_client.build_request(
                method=method,
                url=urljoin(self.server_url, url),
                headers=headers,
                json=json,
                files=files,
                params=params,
            )
            response = await self.httpx_client.send(request, stream=stream)
            if response.status_code == 401:
                raise ShabtiTokenExpiredError(status_code=401)
            if response.status_code == 403:
                raise ShabtiAuthenticationError(status_code=403)
            if response.status_code not in EXPECTED_CODES:
                await response.aread()
                raise_error(response)
            return response

        token = self.token
        task = create_task(make_request(token))

        try:
            if self.refresh_task:
                await self.refresh_task
            result = await task
            return result
        except ShabtiTokenExpiredError:
            # if the token being used is still the same as the stored one, it probably expired
            if token == self.token:
                # if we're not already refreshing, we should launch the refresh task
                if self.refresh_task is None:

                    async def do_refresh():
                        self.token = await self.keycloak_client.a_refresh_token(
                            self.token["refresh_token"]
                        )

                    self.refresh_task = create_task(do_refresh())

                    # unset refresh task when done
                    def on_done(_):
                        self.refresh_task = None

                    self.refresh_task.add_done_callback(on_done)
                # once we've ensured a refresh task is running, wait for it to complete
                await self.refresh_task
            # if not it has probably been refreshed and we're good to go
            # try to rerun using current token
            return await self._make_request(
                method=method, url=url, json=json, files=files, stream=stream
            )

    async def _stream_request(
        self, method, url, json=None, files=None, params: dict = None
    ):
        response = await self._make_request(
            method=method, url=url, json=json, files=files, stream=True, params=params
        )
        async for line in response.aiter_lines():
            yield line
        await response.aclose()

    async def create_collection(
        self, collection_name: str, location: str, owner_username: str | None = None
    ) -> str:
        response = await self._make_request(
            "POST",
            "collections",
            {
                "collection_name": collection_name,
                "location": location,
                "owner_username": owner_username,
            },
        )
        return response.json()["collection_id"]

    async def get_collections(self):
        response = await self._make_request("GET", "collections")
        return [AuthzCollectionInfo(**item) for item in response.json()]
